# 2. Estimación de pi con puntos, usando un círculo.

In [41]:
import numpy as np
import math as mt
import matplotlib.pyplot as plt
#n es el números de puntos.
n = 10000000

Generando "n" valores aleatorios, que vayan del 0 al 1, 2 veces (para "y" y "x"), calcularemos si el punto en cuestión queda dentro del circulo o no. Esto se puede saber creando un circulo imaginario de radio 1, si consideramos el (0,0) como el centro del circulo, la distancia de cualquier parte del borde del circulo al centro siempre será de 1. Teniendo las coordenadas "x" y "y", podemos calcular la distancia del (0,0) hasta (x,y) simplemente usando pitagoras, si da mayor a 1 es por que el punto queda fuera del circulo, si da menor a 1 queda dentro.

El resultado de cantidad de puntos que calleron dentro del circulo se debe dividir con el total de puntos usados, así tendremos un valor mas o menos acercado a pi. Como solo estamos trabajando en el cuadrante 1 de la gráfica, no trabajamos con valores negativos, por lo que el circulo está dividido en 4, para llevarlo a la estimación de pi total, debemos multiplicar por 4.

Esto nos debe dar un número cercano a pi.

In [43]:
rnd = np.random.default_rng(42)
valores_x = rnd.random(n)
valores_y = rnd.random(n)
dentro  = 0
for i in range(n):
    pit = mt.sqrt(valores_y[i]**2+valores_x[i]**2)
    if pit < 1:
        dentro = dentro+1
dentro*4/n

3.141304

Usando 10 millones de puntos distintos, la estimación de pi es correcta los primeros 3 decimales, 3.14130 donde pi real da 3.14159.